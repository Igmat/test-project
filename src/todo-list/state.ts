import { action, computed, observable } from 'mobx';
import TodoService, { ITodo } from './service';

let lastId = 0;
export class Todo implements ITodo {
    id: number;
    @observable isCompleted: boolean;
    @observable name: string;
    constructor(name: string, id?: number, isCompleted?: boolean) {
        this.name = name;
        if (id === undefined) {
            this.id = lastId;
            lastId++;
        } else {
            this.id = id;
        }
        this.isCompleted = !!isCompleted;
    }

    @action complete = async () => {
        this.isCompleted = true;
        return await TodoService.updateTodo(this)
    }
    @action toggle = async () => {
        this.isCompleted = !this.isCompleted;
        return await TodoService.updateTodo(this)
    }
}
export class TodoState {
    @observable lastId: number = 0;
    @observable entities: Todo[] = [];
    @computed get completedAmount() {
        return this.entities.filter(todo => todo.isCompleted).length;
    }
    @computed get uncompletedAmount() {
        return this.entities.filter(todo => !todo.isCompleted).length;
    }
    @computed get totalAmount() {
        return this.entities.length;
    }

    @action add = async (name: string) => {
        const lastId = await TodoService.addTodo({ name, id: this.lastId, isCompleted: false });
        this.entities.push(new Todo(name, lastId));
        this.lastId = lastId;
    }

    constructor() {
        TodoService.todos
            .then(todos =>
                this.entities.push(...todos.map(todo =>
                    new Todo(todo.name, todo.id, todo.isCompleted))));
        TodoService.lastId.then(lastId => this.lastId = lastId);
    }
}

export default new TodoState();