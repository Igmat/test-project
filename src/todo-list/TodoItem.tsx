import { Checkbox, ListItem, ListItemIcon, ListItemText } from '@material-ui/core';
import { observer } from 'mobx-react';
import React from 'react';
import { Todo } from './state';

export interface ITodoItemProps {
    index: number;
    todo: Todo;
}
export const TodoItem = observer(({ index, todo }: ITodoItemProps) =>
    <ListItem key={index} role={undefined} dense button onClick={todo.toggle}>
        <ListItemIcon>
            <Checkbox edge="start" checked={todo.isCompleted} tabIndex={-1} disableRipple inputProps={{ 'aria-labelledby': `checkbox-list-label-${index}` }} />
        </ListItemIcon>
        <ListItemText id={`checkbox-list-label-${index}`} primary={todo.name} />
    </ListItem>);

export default TodoItem;