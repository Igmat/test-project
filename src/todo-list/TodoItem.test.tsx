import React from 'react';
import renderer from 'react-test-renderer';
import { Todo } from './state';
import TodoItem from './TodoItem';

test('App matches snapshot', () => {
  const todo = new Todo('test todo');
  const tree = renderer
    .create(<TodoItem index={0} todo={todo} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});