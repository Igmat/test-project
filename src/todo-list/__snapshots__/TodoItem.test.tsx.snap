// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
<div
  aria-disabled={false}
  className="MuiButtonBase-root MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-button"
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
>
  <div
    className="MuiListItemIcon-root"
  >
    <span
      aria-disabled={false}
      className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-23 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary MuiIconButton-edgeStart"
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={null}
    >
      <span
        className="MuiIconButton-label"
      >
        <input
          aria-labelledby="checkbox-list-label-0"
          checked={false}
          className="PrivateSwitchBase-input-26"
          data-indeterminate={false}
          onChange={[Function]}
          tabIndex={-1}
          type="checkbox"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </span>
    </span>
  </div>
  <div
    className="MuiListItemText-root MuiListItemText-dense"
    id="checkbox-list-label-0"
  >
    <span
      className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
    >
      test todo
    </span>
  </div>
</div>
`;
